akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/TypedSerializer.scala

```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.scaladsl
+
+/**
+ * Used in [[akka.actor.typed.internal.routing.RoutingLogics.ConsistentHashingLogic]].
+ */
+trait TypedSerializer[T] {
+  def apply(message: T): Array[Byte]
+}
```

Comment:
I think this is the way to go, but the name should be more specific and not inclute `typed` because we are already in the `akka.actor.typed` namespace, `RoutingHashExtractor` or something like that.

Since it can be shared between Java and Scala APIs it should live in `akka.actor.typed` rather than `scaladsl`.

Still thinking about if we should have a way to do the `T => String` rather than `T => Array[Byte]` somehow, wdyt, is that not useful enough? If not we should make sure we have a sample of that (even if that is quite straight forward)

Reply:
I think it would be great, but unless we serious type disjunction or add implicit class for conversion, then it will be hard to do something else then what was already introduced in classic part of akka.

I think I will stay with `Array[Byte]` and add proper sample and documentation, in future, when migration to dotty will be in place then union types will be available :)

Example approach with implicit classes:
```
sealed trait RoutingHashExtractor[T] {
  def apply(message: T): HashExtract
}

object RoutingHashExtractor{
  sealed trait HashExtract
  case class StringHashExtract(value: String) extends HashExtract
  case class ArrayByteHashExtract(value: Array[Byte]) extends HashExtract

  implicit class StringOps(a: String) {
    def toHashExtract: HashExtract = StringHashExtract(a)
  }

  implicit class ArrayByteOps(a: Array[Byte]) {
    def toHashExtract: HashExtract = ArrayByteHashExtract(a)
  }
}
```

Example approach with explicit type disjunction
```
sealed trait RoutingHashExtractor[T] {
  
  def apply(message: T): String ∨ Array[Byte]
}

object RoutingHashExtractor {
  // https://milessabin.com/blog/2011/06/09/scala-union-types-curry-howard/
  private type ¬[A] = A => Nothing
  private type ∨[T, U] = ¬[¬[T] with ¬[U]]
}
```

Reply:
I don't think it's worth doing any complicated stuff like that

Perhaps something like
```
// shared between DSLs, is a SAM so can be created from lambda in both Java and Scala 
@FunctionalInterface
trait HashExtractor[T] {
  def apply(m: T): Array[Byte]
}

// in the respective DSLs (to get the different function types right)
object HashExtractor {
  def stringExtractor[T](f: T => String): HashExtractor[T] = ...
}
```

Reply:
That's actually the approach I started implementing, thank you.